src/data/polynomial/erase_lead.lean

```diff
@@ -172,18 +173,43 @@ begin
   induction c with c hc,
   { assume f df f0,
     convert P_0,
-    simpa only [support_eq_empty, card_eq_zero] using f0},
-
- --   exact λ f df f0, by rwa (finsupp.support_eq_empty.mp (card_eq_zero.mp f0)) },
+    simpa only [support_eq_empty, card_eq_zero] using f0 },
   { intros f df f0,
-    rw ← erase_lead_add_C_mul_X_pow f,
-    refine P_C_add f.erase_lead _ (erase_lead_nat_degree_le.trans df) _ _ _,
+    rw [← erase_lead_add_C_mul_X_pow f],
+    cases c,
+    { convert P_C_mul_pow f.nat_degree f.leading_coeff _ df,
+      { convert zero_add _,
+        rw [← card_support_eq_zero, erase_lead_card_support f0] },
+      { rw [leading_coeff_ne_zero, ne.def, ← card_support_eq_zero, f0],
+        exact zero_ne_one.symm } },
+    refine P_C_add f.erase_lead _ _ (erase_lead_nat_degree_le.trans df) _ _ _,
+    { refine (erase_lead_nat_degree_lt _).trans_le (le_of_eq _),
+      { exact (nat.succ_le_succ (nat.succ_le_succ (nat.zero_le _))).trans f0.ge },
+      { rw [nat_degree_C_mul_X_pow _ _ (leading_coeff_ne_zero.mpr _)],
+        rintro rfl,
+        simpa using f0 } },
     { exact (nat_degree_C_mul_X_pow_le f.leading_coeff f.nat_degree).trans df },
     { exact hc _ (erase_lead_nat_degree_le.trans df) (erase_lead_card_support f0) },
     { refine P_C_mul_pow _ _ _ df,
-      rw [ne.def, leading_coeff_eq_zero],
-      rintro rfl,
-      exact not_le.mpr c.succ_pos f0.ge } }
+      rw [ne.def, leading_coeff_eq_zero, ← card_support_eq_zero, f0],
+      exact nat.succ_ne_zero _ } }
+end
+
+lemma map_nat_degree_eq_nat_degree (F : polynomial R → polynomial R) (p)
+  (F0 : F 0 = 0)
+  (F_add : ∀ p q, F (p + q) = F p + F q)
```

Comment:
Is there any reason to

- avoid `polynomial R →+ polynomial R` or even `{f : F} [add_monoid_hom_class F (polynomial R) (polynomial R)]`?
- use `polynomial R → polynomial R` instead of `polynomial R → polynomial S`?

Reply:
Thank you very much for these suggestions!  I went for the `add_monoid_hom_class` version to `polynomial S`.